Braze-Demo-LAB-Register-Content-Extension/NotificationViewController.swift

```diff
@@ -0,0 +1,141 @@
+    let userInfo = content.userInfo
+
+    titleLabel.text = userInfo[PushNotificationKey.certificationTitle.rawValue] as? String
+    descriptionLabel.text = userInfo[PushNotificationKey.certificationDescription.rawValue] as? String
+    imageView.image = imageFromNotification(content: content)
+  }
+
+  func didReceive(_ response: UNNotificationResponse, completionHandler completion: @escaping (UNNotificationContentExtensionResponseOption) -> Void) {
+    if response.actionIdentifier == registerIdentifier {
+      UINotificationFeedbackGenerator().notificationOccurred(.success)
+        
+      extensionContext?.notificationActions = []
+      saveRegisteredForCertificationEvent()
+      saveEmailAttribute()
+      
+      displayAllSetView { completion(.dismiss) }
+    } else {
+      completion(.doNotDismiss)
+    }
+  }
+}
+
+// MARK: - Email TextField Delegate
+extension NotificationViewController: UITextFieldDelegate {
+  func textFieldDidEndEditing(_ textField: UITextField) {
+    setRegisterEmail(textField.text)
+  }
+  
+  func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+    return textField.resignFirstResponder()
+  }
+}
+
+// MARK: - Private Methods
+private extension NotificationViewController {
+  /// Loading an image from a push notifcation.
+  ///
+  /// The `Service Extension` target must be included with the `ContentExtension` target
+  func imageFromNotification(content: UNNotificationContent) -> UIImage? {
+    guard let attachment = content.attachments.first else { return nil }
+    guard attachment.url.startAccessingSecurityScopedResource() else { return nil }
+    let data = try? Data(contentsOf: attachment.url)
+    attachment.url.stopAccessingSecurityScopedResource()
+    
+    return data != nil ? UIImage(data: data!) : nil
+  }
+  
+  func setRegisterEmail(_ email: String?) {
+    if let email = email, email.isValidEmail {
+      registerEmail = email
+    } else {
+      registerEmail = nil
+    }
+  }
+  
+  func displayAllSetView(completion: @escaping () -> ()) {
+    let allSetView: AllSetView = .fromNib()
+    allSetView.frame = view.frame
+    allSetView.alpha = 0.0
+    view.addSubview(allSetView)
+    
+    UIView.animate(withDuration: 1.0) {
+      allSetView.alpha = 1.0
+    } completion: { _ in
+      DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
+        completion()
+      }
+    }
+  }
+}
+
+// MARK: - Analytics
+private extension NotificationViewController {
+  /// Saves a custom event to `userDefaults` with the given suite name that is your `App Group` name.  The value `"Event Name`" is explicity saved.
+  ///
+  /// There is a conditional unwrap to check if there are saved pending events (in the case of multiple registrations) and appends the event or saves a new array with one event.
+  func saveRegisteredForCertificationEvent() {
+    let customEventDictionary = [["Event Name": "Registered for Certification"]] as [[String : Any]]
```

Comment:
Is `Event Name` stored for every custom event? Any way to avoid that duplication?